### PR TITLE
improve ReportActionItem performance

### PR DESCRIPTION
**Pull request by OlimpiaZurek**
### Explanation of Change

This PR moves the useOnyx connections from the ReportActionItem level up to the list level.

### Fixed Issues

$ https://github.com/Expensify/App/issues/66149
PROPOSAL:

### Tests

1. Sign up to the app.
2. Open any report.
3. Verify that all messages and their elements are displayed correctly in the chat.
4. Post a new message and react with an emoji.
5. Verify that the emoji is displayed correctly.
6. Press "Edit comment".
7. Verify that the edited message is displayed properly.

8. Open a report that contains Expenses.
9. Click on the Expense Report item.
10. Verify that the expenses list and all its elements (e.g., date, merchant, total) are displayed correctly.
11. Add a new message to the expenses list.
12. Repeat steps 5–7 for this message.

13. Create duplicated expenses.
14. Check if the errors on expense items are displayed correctly.
15. Create an expense with  eg. missing category.
16. Verify that the correct error is shown.

17. Navigate to the Reports section by tapping the bottom tab.
18. Open the "Search Type" menu and switch to "Chats".
19. Confirm that a list of chat reports is displayed properly.

- [ ] Verify that no errors appear in the JS console

### QA Steps

// TODO: These must be filled out, or the issue title must include "[No QA]."

Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I verified there are no new alerts related to the `canBeMissing` param for `useOnyx`
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L80)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L116-L123)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

[android_native.webm](https://github.com/user-attachments/assets/2b21db90-8ffb-491a-99af-40e93a4c21ce)

</details>

<details>
<summary>Android: mWeb Chrome</summary>

[android_mweb.webm](https://github.com/user-attachments/assets/ccaf14d2-1d0b-4428-b6ce-d690349db42e)

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/5dff623c-62c1-4fe4-8007-72376791a7be

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/0d75f2c9-1ac0-4a87-98d3-922d6429d242

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/5d0cb7b1-4c57-45a6-9fb9-45426b4a2c3e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/2624af08-f2ed-4eac-aabd-5d6b5b0a0ddc

</details>

### CODE REVIEW

**Comment by OlimpiaZurek** [comment]
=== 📁 Trace File Analysis ===

📄 File: before.json
  • Commit count: 36
  • Total render duration: 2238.065 ms
  • p50 / p95 / p99 (ms):
    • p50: 12.102 ms
    • p95: 178.171 ms
    • p99: 992.416 ms

📄 File: after.json
  • Commit count: 29
  • Total render duration: 1824.265 ms
  • p50 / p95 / p99 (ms):
    • p50: 4.234 ms
    • p95: 325.231 ms
    • p99: 507.985 ms

=== 📊 Trace Comparison ===

🔁 Comparing Baseline ↔️ after.json
  🟢 Duration Diff: -413.800 ms
  🟢 Commit Count Diff: -7

File: src/pages/Debug/ReportAction/DebugReportActionCreatePage.tsx [hunk 1]
```diff
@@ -120,6 +124,13 @@ function DebugReportActionCreatePage({
                                     index={0}
                                     isFirstVisibleReportAction={false}
                                     shouldDisplayContextMenu={false}
+                                    userWallet={userWallet}
+                                    isUserValidated={isUserValidated}
+                                    personalDetails={personalDetailsList}
+                                    userBillingFundID={userBillingFundID}
+                                    draftMessage={undefined}
+                                    emojiReactions={undefined}
+                                    linkedTransactionRouteError={undefined}
```
Comment (hoangzinh) [inline src/pages/Debug/ReportAction/DebugReportActionCreatePage.tsx]: Aren't they `undefined` by default?

File: src/pages/Debug/ReportAction/DebugReportActionPreview.tsx [hunk 2]
```diff
@@ -17,6 +17,10 @@ type DebugReportActionPreviewProps = {
 function DebugReportActionPreview({reportAction, reportID}: DebugReportActionPreviewProps) {
     const [allReports] = useOnyx(ONYXKEYS.COLLECTION.REPORT, {canBeMissing: false});
     const [policies] = useOnyx(ONYXKEYS.COLLECTION.POLICY, {canBeMissing: false});
+    const [userWallet] = useOnyx(ONYXKEYS.USER_WALLET, {canBeMissing: true});
+    const [isUserValidated] = useOnyx(ONYXKEYS.ACCOUNT, {selector: (account) => account?.validated, canBeMissing: true});
+    const [personalDetails] = useOnyx(ONYXKEYS.PERSONAL_DETAILS_LIST, {canBeMissing: true});
```
Comment (hoangzinh) [inline src/pages/Debug/ReportAction/DebugReportActionPreview.tsx]: Let's get from OnyxProvider here `const personalDetails = usePersonalDetails();`

File: src/pages/TransactionDuplicate/DuplicateTransactionItem.tsx [hunk 3]
```diff
@@ -19,6 +20,10 @@ type DuplicateTransactionItemProps = {
 
 function DuplicateTransactionItem({transaction, index, allReports, policies}: DuplicateTransactionItemProps) {
     const styles = useThemeStyles();
+    const [userWallet] = useOnyx(ONYXKEYS.USER_WALLET, {canBeMissing: true});
+    const [isUserValidated] = useOnyx(ONYXKEYS.ACCOUNT, {selector: (account) => account?.validated, canBeMissing: true});
+    const [personalDetails] = useOnyx(ONYXKEYS.PERSONAL_DETAILS_LIST, {canBeMissing: true});
```
Comment (hoangzinh) [inline src/pages/TransactionDuplicate/DuplicateTransactionItem.tsx]: Same here `const personalDetails = usePersonalDetails();`

File: src/pages/home/report/ReportActionsListItemRenderer.tsx [hunk 4]
```diff
@@ -61,6 +62,27 @@ type ReportActionsListItemRendererProps = {
...
+    linkedTransactionRouteError?: OnyxEntry<Errors>;
+
+    /** Whether the user is validated */
+    isUserValidated?: boolean | undefined;
+
+    /** Personal details list */
+    personalDetails?: OnyxEntry<PersonalDetailsList>;
+
+    /** User billing fund ID */
+    userBillingFundID?: number | undefined;
```
Comment (hoangzinh) [inline src/pages/home/report/ReportActionsListItemRenderer.tsx]: Those new props should be required props. Otherwise, if someone uses this component to render and forgets to pass those props, it will render incorrectly.

File: src/pages/home/report/ReportActionsList.tsx [hunk 5]
```diff
@@ -600,25 +615,38 @@ function ReportActionsList({
                     isFirstVisibleReportAction={firstVisibleReportActionID === reportAction.reportActionID}
                     shouldUseThreadDividerLine={shouldUseThreadDividerLine}
                     transactions={Object.values(transactions ?? {})}
+                    userWallet={userWallet}
+                    isUserValidated={isUserValidated}
+                    personalDetails={personalDetailsList}
+                    draftMessage={reportDraftMessages as OnyxEntry<string & OnyxTypes.ReportActionsDrafts>}
```
Comment (hoangzinh) [inline src/pages/home/report/ReportActionsList.tsx]: We can pass draftMessages of reportAction here, instead of the whole draftMessages of a report. What do you think?

File: src/pages/Debug/ReportAction/DebugReportActionCreatePage.tsx (hunk 1 above)
Comment (OlimpiaZurek) [reply src/pages/Debug/ReportAction/DebugReportActionCreatePage.tsx]: Yes, removed

File: src/pages/Debug/ReportAction/DebugReportActionPreview.tsx (hunk 2 above)
Comment (OlimpiaZurek) [reply src/pages/Debug/ReportAction/DebugReportActionPreview.tsx]: Changed

File: src/pages/TransactionDuplicate/DuplicateTransactionItem.tsx (hunk 3 above)
Comment (OlimpiaZurek) [reply src/pages/TransactionDuplicate/DuplicateTransactionItem.tsx]: Changed

File: src/pages/home/report/ReportActionsListItemRenderer.tsx (hunk 4 above)
Comment (OlimpiaZurek) [reply src/pages/home/report/ReportActionsListItemRenderer.tsx]: Done, but I left a few properties as not required since they aren't needed in every component that renders `ReportActionItem`

File: src/pages/home/report/ReportActionsList.tsx (hunk 5 above)
Comment (OlimpiaZurek) [reply src/pages/home/report/ReportActionsList.tsx]: Already updated it as suggested

Comment (hoangzinh) [reply src/pages/home/report/ReportActionsList.tsx]: Thank you, @OlimpiaZurek. Let me take a 2nd round

File: src/pages/home/report/ReportActionsListItemRenderer.tsx (hunk 4 above)
Comment (hoangzinh) [reply src/pages/home/report/ReportActionsListItemRenderer.tsx]: Hmm. Then it's easy to cause a bug like this, the `emojiReactions` are not displayed in `MoneyRequestReportView`. Can you also check for `linkedTransactionRouteError`?

Screenshot 2025-07-24 at 17 22 29

File: src/pages/home/report/ReportActionItemParentAction.tsx [hunk 6]
```diff
@@ -60,6 +61,27 @@ type ReportActionItemParentActionProps = {
...
     shouldUseThreadDividerLine?: boolean;
+
+    /** User wallet */
+    userWallet?: OnyxEntry<OnyxTypes.UserWallet>;
+
+    /** Whether the user is validated */
+    isUserValidated?: boolean | undefined;
+
+    /** Personal details list */
+    personalDetails?: OnyxEntry<OnyxTypes.PersonalDetailsList>;
```
Comment (hoangzinh) [inline src/pages/home/report/ReportActionItemParentAction.tsx]: should we declare those props as required props

File: src/pages/home/report/ReportActionsListItemRenderer.tsx (hunk 4 above)
Comment (OlimpiaZurek) [reply src/pages/home/report/ReportActionsListItemRenderer.tsx]: Checked `MoneyRequestReportActionsList` and fixed the emoji and draftMessage props — turns out both are needed there.

As for `linkedTransactionRouteError`, it looks like we don’t need it in this component.

Here are screenshots from staging:
staging

And here’s one from the current solution
current

Comment (hoangzinh) [reply src/pages/home/report/ReportActionsListItemRenderer.tsx]: Thank you @OlimpiaZurek

**Review by hoangzinh — APPROVED** [review]
LGTM

**Review by mountiny — APPROVED** [review]
Changes look good to me but i think we should add much more manual tests steps, for example the duplicates flow and others that are touched by these changes

**Comment by OlimpiaZurek** [comment]
@mountiny Added a few more test steps. I think they should cover all cases touched by these changes

File: src/components/MoneyRequestReportView/MoneyRequestReportActionsList.tsx [hunk 7]
```diff
@@ -471,6 +479,12 @@ function MoneyRequestReportActionsList({
                 !isConsecutiveChronosAutomaticTimerAction(visibleReportActions, index, chatIncludesChronosWithID(reportAction?.reportID)) &&
                 hasNextActionMadeBySameActor(visibleReportActions, index);
 
+            const actionEmojiReactions = emojiReactions?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_REACTIONS}${reportAction.reportActionID}`];
+            const originalReportID = getOriginalReportID(report.reportID, reportAction);
+            const reportDraftMessages = draftMessage?.[`${ONYXKEYS.COLLECTION.REPORT_ACTIONS_DRAFTS}${originalReportID}`];
```
Comment (NikkiWines) [inline src/components/MoneyRequestReportView/MoneyRequestReportActionsList.tsx:484]: @OlimpiaZurek I think making draft loading dependent on the `getOriginalReportID`  is causing [this deploy blocker](https://github.com/Expensify/App/issues/67148). 

BUT, it seems like it's because the logic in `getOriginalReportID` was already slightly flawed, and this PR just happened to surface the existing bug.

In `getOriginalReportID`: 

```
const isThreadReportParentAction = reportAction?.childReportID?.toString() === reportID;
if (Object.keys(currentReportAction ?? {}).length === 0) {
    return isThreadReportParentAction ? getReport(reportID, allReports)?.parentReportID : (transactionThreadReportID ?? reportID);
}
```

If `currentReportAction` is empty the `isThreadReportParentAction` condition incorrectly evaluates to `true` and the function incorrectly returns the `parentReportID` instead of the `transactionThreadReportID`. 

And we didn't see it before because each component had its own draft loading logic.
